beginner_source/onnx/onnx_registry_tutorial.py

```diff
@@ -1,303 +1,247 @@
+#
+# * Overriding the implementation of an existing PyTorch operator
+# * Using custom ONNX operators
+# * Supporting a custom PyTorch operator
+#
+# Overriding the implementation of an existing PyTorch operator
+# -------------------------------------------------------------
+#
+# Although the ONNX exporter team does their best efforts to support all PyTorch operators, some of them
+# might not be supported yet. In this section, we will demonstrate how you can add
+# unsupported PyTorch operators to the ONNX Registry.
 #
-# During the model export to ONNX, the PyTorch model is lowered to an intermediate
-# representation composed of `ATen operators <https://pytorch.org/docs/stable/torch.compiler_ir.html>`_.
-# While ATen operators are maintained by PyTorch core team, it is the responsibility of the ONNX exporter team
-# to independently implement each of these operators to ONNX through `ONNX Script <https://onnxscript.ai/>`_.
-# The users can also replace the behavior implemented by the ONNX exporter team with their own implementation
-# to fix bugs or improve performance for a specific ONNX runtime.
+# .. note::
+#       The steps to implement unsupported PyTorch operators are the same to replace the implementation of an existing
+#       PyTorch operator with a custom implementation.
+#       Because we don't actually have an unsupported PyTorch operator to use in this tutorial, we are going to leverage
+#       this and replace the implementation of ``torch.ops.aten.add.Tensor`` with a custom implementation the same way we would
+#       if the operator was not implemented by the ONNX exporter.
+#
+# When a model cannot be exported to ONNX due to an unsupported operator, the ONNX exporter will show an error message
+# similar to:
+#
+# .. code-block:: python
 #
-# The ONNX Registry manages the mapping between PyTorch operators and the ONNX operators counterparts and provides
-# APIs to extend the registry.
+#   No decompositions registered for [...]
 #
-# In this tutorial, we will cover three scenarios that require extending the ONNX registry with custom operators:
+# The error message indicates that the unsupported PyTorch operator is ``torch.ops.aten.add.Tensor``.
+# The operator is of type ``<class 'torch._ops.OpOverload'>``, and this operator is what we will use as the
+# target to register our custom implementation.
 #
-# * Custom operators with existing ONNX Runtime support
-# * Custom operators without ONNX Runtime support
+# To add support for an unsupported PyTorch operator or to replace the implementation for an existing one, we need:
 #
+# * The target PyTorch operator.
+# * The implementation of the operator using `ONNX Script <https://github.com/microsoft/onnxscript>`__.
+#   ONNX Script is a prerequisite for this tutorial. Please make sure you have read the
+#   `ONNX Script tutorial <https://github.com/microsoft/onnxscript/blob/main/docs/tutorial/index.md>`_
+#   before proceeding.
```

Comment:
I suggest you add this whole list to a Prerequisites section.

Reply:
Done